### ISSUE TITLE
how to get private and public key with hexstring format

### ISSUE THREAD

**Issue by kimi20071025** [openssl/openssl#11245]

**Comment by kimi20071025** [comment]
I user the following code to get a EVP_PKEY pointer
    ```
	EVP_PKEY *pkey = NULL;
	EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(NID_X25519, NULL);
	EVP_PKEY_keygen_init(pctx);
	EVP_PKEY_keygen(pctx, &pkey);
	EVP_PKEY_CTX_free(pctx);
    ```
I know I can get the PEM format private key and public key by "PEM_write_PUBKEY" and "PEM_write_PrivateKey". the result is like this:
public key:"MCowBQYDK2VuAyEA29XWmtbvyjfIzO9JtOIcMy44hBloizCOT5y4k+b0DEw="

I would like to known how can I get the HEXSTRING format like this:
"c53c22208b61860b06c62e5406a7b330c2b577aa5558981510d128247d38bd1d"

**Comment by beldmit** [comment]
The variant 
```public key:"MCowBQYDK2VuAyEA29XWmtbvyjfIzO9JtOIcMy44hBloizCOT5y4k+b0DEw="```
is base64 encoding. If there are functions DER_write_PrivateKey/DER_write_PUBKEY, you can get them in binary form, otherwise, you can decode it from base64 and then use `OPENSSL_buf2hexstr`.

**Comment by kimi20071025** [comment]
> The variant
> `public key:"MCowBQYDK2VuAyEA29XWmtbvyjfIzO9JtOIcMy44hBloizCOT5y4k+b0DEw="`
> is base64 encoding. If there are functions DER_write_PrivateKey/DER_write_PUBKEY, you can get them in binary form, otherwise, you can decode it from base64 and then use `OPENSSL_buf2hexstr`.

Maybe I had not describle my problem in detail. What I need is original private key and public key, Not the concealed one. 
Are there any APIs to get the original private key and public key directly by EVP__PKEY? 
The functions EVP_PKEY_get_raw_private_key()/EVP_PKEY_get_raw_public_key() are what I need?

**Comment by mattcaswell** [comment]
> The functions EVP_PKEY_get_raw_private_key()/EVP_PKEY_get_raw_public_key() are what I need?

It kind of depends what you mean by the "original private key and public key, Not the concealed one". Those functions get the raw X25519 private key and public key in binary format. This is probably what you meant. *Possibly* you mean the DER encoding of the X25519 key which can be obtained with `i2d_PUBKEY` and `i2d_PrivateKey`

**Comment by kimi20071025** [comment]
> > The functions EVP_PKEY_get_raw_private_key()/EVP_PKEY_get_raw_public_key() are what I need?
> 
> It kind of depends what you mean by the "original private key and public key, Not the concealed one". Those functions get the raw X25519 private key and public key in binary format. This is probably what you meant. _Possibly_ you mean the DER encoding of the X25519 key which can be obtained with `i2d_PUBKEY` and `i2d_PrivateKey`

Thank you for your reply, the raw  private key and public key is what I need. 
But when I call the function EVP_PKEY_get_raw_private_key like is, it returns fail, why?
Do I call it in a wrong way?

int main() {
	
	/* Generate private and public key */
	EVP_PKEY *pkey = NULL;
	EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_X25519, NULL);
	EVP_PKEY_keygen_init(pctx);
	EVP_PKEY_keygen(pctx, &pkey);
	EVP_PKEY_CTX_free(pctx);	
	
	char privateKeyRaw[32] = {0};
	size_t privateLen = 0;
	if (1 != EVP_PKEY_get_raw_private_key(pkey, (unsigned char*)privateKeyRaw, &privateLen))	
	{
		std::cout<<"err, len is "<<privateLen<<std::endl;		
	}
        return 0;
}

**Comment by mattcaswell** [comment]
Ah! This is actually a documentation bug. You are supposed to populate `privateLen` with the size of the `privateKeyRaw` buffer on entry to `EVP_PKEY_get_raw_private_key()`, i.e. making this change should fix it for you:

````
size_t privateLen = sizeof(privateKeyRaw);
````

You could be forgiven for not knowing this however since the docs make no mention of it. Fix for that in #11254.

**Comment by kimi20071025** [comment]
> Ah! This is actually a documentation bug. You are supposed to populate `privateLen` with the size of the `privateKeyRaw` buffer on entry to `EVP_PKEY_get_raw_private_key()`, i.e. making this change should fix it for you:
> 
> ```
> size_t privateLen = sizeof(privateKeyRaw);
> ```
> 
> You could be forgiven for not knowing this however since the docs make no mention of it. Fix for that in #11254.

thank you, I solved the problem according to your advise.
